File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.py
class Solution:

    def __init__(self) -> None:
        self.m:int =0
        self.n:int =0
        self.reach_cache = [[]]
        self.cal = [[]]

    class Reach:
        def __init__(self) -> None:
            self.atlantic = False
            self.pacific = False
    
    def bfs(self, heights, i:int, j:int) -> Reach:
        visited = set()
        queue:Deque[Tuple[int]] = deque()
        reach:Solution.Reach = Solution.Reach()

        def addCell(x:int, y:int, par_height:int):
            if x < 0 or y < 0:
                reach.pacific = True
                return
            if y >= self.n or x >= self.m:
                reach.atlantic = True
                return

            if (par_height != -1 and par_height < heights[x][y]):
                return

            if self.cal[x][y] == True:
                reach.pacific = self.reach_cache[x][y].pacific or reach.pacific
                reach.atlantic = self.reach_cache[x][y].atlantic or reach.atlantic
                return

            if not (x,y) in visited:
                queue.append((x,y))

        addCell(i,j, -1)
        while queue:
            (x,y) = queue.popleft()
            visited.add((x,y))
            addCell(x+1, y, heights[x][y])
            addCell(x, y+1, heights[x][y])
            addCell(x-1, y, heights[x][y])
            addCell(x, y-1, heights[x][y])
        self.reach_cache[i][j] = reach
        self.cal[i][j] = True
        return reach



    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        sol = []
        self.m = len(heights)
        self.n = len(heights[0])
        self.reach_cache = [[self.Reach()] * self.n for _ in range(self.m)]
        self.cal = [[False] * self.n for _ in range(self.m)]
        for i in range(self.m):
            for j in range(self.n):
                reach = self.bfs(heights,i,j)
                if reach.atlantic == True and reach.pacific == True:
                    sol.append([i,j])
        return sol

File: 417-pacific-atlantic-water-flow/pacific-atlantic-water-flow.cpp
class Solution {
public:
    int get_height(vector<vector<int>>& heights, int x, int y) {
        int n = heights.size();
        int m = heights[0].size();
        if(x < 0 || x >= n|| y < 0 || y >= m) return -1;
        return heights[x][y];

    }

    bool is_valid(vector<vector<int>>& heights,vector<vector<bool>>& visited, int x, int y, int comp_height){
        int n = heights.size();
        int m = heights[0].size();
        if(x < 0 || x >= n|| y < 0 || y >= m) return false;
        if(visited[x][y]) return false;
        if(heights[x][y] < comp_height) return false;
        return true;
    }

    void bfs(vector<vector<int>>& heights, vector<vector<bool>>& visited, queue<pair<int, int>>& q){
        while(!q.empty()){
            pair<int, int> coords = q.front();q.pop();
            vector<pair<int, int>> directions({{1, 0}, {-1, 0}, {0, 1}, {0, -1}});
            for(const auto& dir: directions) {
                int new_x = coords.first + dir.first;
                int new_y = coords.second + dir.second;
                if(is_valid(heights, visited,new_x, new_y, get_height(heights, coords.first, coords.second))){
                    visited[new_x][new_y] = true;
                    q.push({new_x, new_y});
                }
            }
        }
    }
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int n = heights.size();
        int m = heights[0].size();
        vector<vector<bool>> atlantic(n, vector<bool>(m, false)), pacific(n, vector<bool> (m, false));
        queue<pair<int, int>> atlantic_q, pacific_q;
        for(int i = 0; i < n; i++){
            atlantic_q.push({i, -1});
            pacific_q.push({i, m});
        }
        for(int j = 0; j < m; j++){
            atlantic_q.push({-1, j});
            pacific_q.push({n, j});
        }
        bfs(heights, atlantic, atlantic_q);
        bfs(heights, pacific, pacific_q);
        vector<vector<int>> sol;
        for(int i = 0 ; i< n ; i++){
            for(int j = 0; j < m; j++){
                if(atlantic[i][j] && pacific[i][j]){
                    sol.push_back({i,j});
                }
            }
        }
        return sol;
    }
};